File: backend/services/workflow_manager.py
"""
==========================================================
Easy AI Studio
File    : backend/services/workflow_manager.py
Version : 1.0.0
Sprint  : 10
==========================================================
"""

from datetime import datetime
from typing import Dict
from typing import List
from typing import Optional
from uuid import uuid4


class WorkflowManager:

    def __init__(self):

        self.workflows: Dict[str, dict] = {}

    def create(

        self,

        name: str,

        nodes: list = None,

        connections: list = None

    ) -> dict:

        workflow = {

            "id": str(uuid4()),

            "name": name,

            "nodes": nodes or [],

            "connections": connections or [],

            "enabled": True,

            "created_at": datetime.now().isoformat(),

            "updated_at": datetime.now().isoformat()

        }

        self.workflows[workflow["id"]] = workflow

        return workflow

    def get(

        self,

        workflow_id: str

    ) -> Optional[dict]:

        return self.workflows.get(workflow_id)

    def all(self) -> List[dict]:

        return list(self.workflows.values())

    def update(

        self,

        workflow_id: str,

        data: dict

    ) -> Optional[dict]:

        workflow = self.workflows.get(workflow_id)

        if workflow is None:

            return None

        workflow.update(data)

        workflow["updated_at"] = datetime.now().isoformat()

        return workflow

    def delete(

        self,

        workflow_id: str

    ) -> bool:

        if workflow_id in self.workflows:

            del self.workflows[workflow_id]

            return True

        return False

    def enable(

        self,

        workflow_id: str

    ) -> bool:

        workflow = self.get(workflow_id)

        if workflow:

            workflow["enabled"] = True

            return True

        return False

    def disable(

        self,

        workflow_id: str

    ) -> bool:

        workflow = self.get(workflow_id)

        if workflow:

            workflow["enabled"] = False

            return True

        return False

    def count(self) -> int:

        return len(self.workflows)

    def clear(self):

        self.workflows.clear()


workflow_manager = WorkflowManager()

File: frontend_old/js/workflowManager.js
/* ==========================================================
   Easy AI Studio
   File: frontend/js/workflowManager.js
   Version: 1.0.0
   Sprint: 6 - Workflow Manager
   ========================================================== */

"use strict";

/* ==========================================================
   Workflow Manager
   ========================================================== */

const WorkflowManager = {

    storageKey: "workflows",

    workflows: [],

    activeWorkflow: null,

    /* ======================================================
       Initialize
       ====================================================== */

    init() {

        this.load();

        console.log(

            "Workflow Manager Ready"

        );

    },

    /* ======================================================
       Load
       ====================================================== */

    load() {

        this.workflows = Storage.get(

            this.storageKey,

            []

        );

    },

    /* ======================================================
       Save
       ====================================================== */

    save() {

        Storage.set(

            this.storageKey,

            this.workflows

        );

    },

    /* ======================================================
       Create Workflow
       ====================================================== */

    create(data = {}) {

        const workflow = {

            id: Utils.uuid(),

            name: data.name || "New Workflow",

            description: data.description || "",

            nodes: [],

            connections: [],

            variables: {},

            enabled: true,

            created: Utils.now(),

            updated: Utils.now()

        };

        this.workflows.push(workflow);

        this.save();

        Notification.success(

            "Workflow created."

        );

        return workflow;

    },

    /* ======================================================
       Open Workflow
       ====================================================== */

    open(id) {

        const workflow = this.get(id);

        if (!workflow) {

            return;

        }

        this.activeWorkflow = id;

        EventBus.emit(

            "workflow.opened",

            workflow

        );

    },

    /* ======================================================
       Add Node
       ====================================================== */

    addNode(workflowId, node) {

        const workflow = this.get(workflowId);

        if (!workflow) {

            return;

        }

        workflow.nodes.push({

            id: Utils.uuid(),

            ...node

        });

        workflow.updated = Utils.now();

        this.save();

    },

    /* ======================================================
       Remove Node
       ====================================================== */

    removeNode(workflowId, nodeId) {

        const workflow = this.get(workflowId);

        if (!workflow) {

            return;

        }

        workflow.nodes = workflow.nodes.filter(

            node => node.id !== nodeId

        );

        workflow.updated = Utils.now();

        this.save();

    },

    /* ======================================================
       Find Workflow
       ====================================================== */

    get(id) {

        return this.workflows.find(

            workflow =>

            workflow.id === id

        );

    },

    /* ======================================================
       Delete Workflow
       ====================================================== */

    remove(id) {

        this.workflows = this.workflows.filter(

            workflow =>

            workflow.id !== id

        );

        if (

            this.activeWorkflow === id

        ) {

            this.activeWorkflow = null;

        }

        this.save();

    },

    /* ======================================================
       Get All
       ====================================================== */

    all() {

        return this.workflows;

    }

};

/* ==========================================================
   Global
   ========================================================== */

window.WorkflowManager = WorkflowManager;